doc/src/documentation_style_guide.rst

```diff
@@ -0,0 +1,1019 @@
+===============================
+SymPy Documentation Style Guide
+===============================
+
+**A Writing Resource for Documentation and Docstrings**
+
+.. contents:: Table Of Contents
+
+General Guidelines
+==================
+
+Documentation is one of the most highly valued aspects of an open source
+project. Documentation teaches users and contributors how to use a project, how
+to contribute, and the standards of conduct within an open source community.
+But according to GitHub’s `Open Source Survey
+<https://opensourcesurvey.org/2017/>`_, incomplete or confusing documentation is
+the most commonly encountered problem in open source. This style guide aims to
+change that.
+
+The purpose of this style guide is to provide the SymPy community with a set of
+style and formatting guidelines that can be utilized and followed when writing
+SymPy documentation. Adhering to the guidelines offered in this style guide
+will bring greater consistency and clarity to SymPy’s documentation, supporting
+its mission to become a full-featured, open source computer algebra system
+(CAS).
+
+The SymPy documentation found at `docs.sympy.org
+<https://docs.sympy.org/latest/index.html>`_ is generated from docstrings in the
+source code and dedicated narrative documentation files in the `doc/src
+directory <https://github.com/sympy/sympy/tree/master/doc/src>`_. Both are
+written in `reStructuredText <http://docutils.sourceforge.net/rst.html>`_ format
+extended by `Sphinx <http://www.sphinx-doc.org/en/master/>`_. Sphinx is also
+used to compile documentation to various formats such as HTML and LaTeX.
```

Comment:
We should list what sphinx extensions we use here also to be complete. The RestructuredText specifciaitons, Sphinx's adiditonal RST specs and features, and any extensions we install alongside sphinx define a based standard specification for supported docstring formatting.

Something like:

The SymPy documentation is parsed by RestructuredText, Sphinx, and the following Sphinx extensions. In addition, SymPy has these modifications that affect how the documentation is parsed. One must write documentation in a format that is collectively defined by these tools, otherwise errors in parsing will occur.

Reply:
We now support the matplotlib extension that makes plots in the docstrings. It would be great to have explanation on how to use them. It is a be tricky I found.

Reply:
Here is my suggested addition for somewhere in this document:

The documentation contained in the `doc/` directory and the docstrings embedded in the Python source code is processed by Sphinx and various Sphinx extensions. This means that the documentation source format is specified by the documentation processing tools. The SymPy Style Guide provides both the essential elements for writing SymPy documentation as wells as any deviations in style we specify relative to these documentation processing tools. The following lists the processing tools:

- ReStructuredText: Narrative documentation files and documentation strings embedded in Python code follow the ReStructuredText format. Advanced features not described in this document can be found at http://docutils.sourceforge.net/rst.html.
- Sphinx: Sphinx includes additional default features for the ReStructuredText specification that are described at: http://www.sphinx-doc.org/
- Sphinx extensions included with Sphinx that we enable:
  - 'sphinx.ext.autodoc': Processes Python source code files for the associated documentation strings to automatically generate pages containing the Application Programming Interface (API). See section on calling autodoc directives in this document to get started. More information is at: https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html
  - 'sphinx.ext.graphviz': Provides a directive for adding Graphviz graphics. See https://www.sphinx-doc.org/en/master/usage/extensions/graphviz.html for more info.
  -  'sphinx.ext.mathjax': Causes math written in LaTeX to display using MathJax in the HTML version of the documentation. More information is at: https://www.sphinx-doc.org/en/master/usage/extensions/math.html#module-sphinx.ext.mathjax. *No bearing on documentation source format*.
  - 'sphinx.ext.linkcode': Causes links to source code to direct to the related files on Github. More information is at: https://www.sphinx-doc.org/en/master/usage/extensions/linkcode.html  *No bearing on documentation source format*.
- Sphinx extensions that are not included with Sphinx that we enable:
  - 'numpydoc': Processes docstrings written in the "numpydoc" format, see https://numpydoc.readthedocs.io. We recommend the subset of numpydoc formatting features in this document. (Note that we currently use an older modified fork of numpydoc, which is included in the SymPy source code).
  - 'sphinx_math_dollar': Allows math to be delimited with dollar signs instead of ReStructuredText directives, e.g. `$a^$` instead of `:math:`a^2``. See https://www.sympy.org/sphinx-math-dollar/ for more info.
  - 'matplotlib.sphinxext.plot_directive': Provides directives for included matplotlib generated figures in ReStructuredText. See https://matplotlib.org/devel/plot_directive.html for more info.
  - 'sympylive': Adds a button on each example in the HTML documentation that opens the example in SymPy Live.  *No bearing on documentation source format*.                     

Everything supported by the above processing tools is available for use in the SymPy documentation, but this styleguide supersedes any recommendations made in the above documents. Note that we do not follow PEP 257 or the www.python.org documentation recommendations.

Reply:
Thanks so much for writing this up @moorepants! I think it makes sense to include this upfront in the General Guidelines, right where you've put it. I will add it!